man/sketchybar.1.scd
sketchybar(1)

# NAME

sketchybar - a highly customizable macOS status bar replacement

# SYNOPSIS

*sketchybar* ++
*sketchybar --config* _FILE_ ++
*sketchybar --reload* [_CONFIG_FILE_] ++
*sketchybar --hotload* _BOOLEAN_

*sketchybar --bar* _SETTING_=_VALUE_ ... _SETTING_=_VALUE_

*sketchybar --add item* _NAME_ _POSITION_ ++
*sketchybar --set* _NAME_ _PROPERTY_=_VALUE_ ... _PROPERTY_=_VALUE_ ++
*sketchybar --default* _PROPERTY_=_VALUE_ ... _PROPERTY_=_VALUE_ ++
*sketchybar --reorder* _NAME_ ... _NAME_ ++
*sketchybar --move* _NAME_ *before*|*after* _REFERENCE_NAME_ ++
*sketchybar --clone* _PARENT_NAME_ _NAME_ [*before*|*after*] ++
*sketchybar --rename* _OLD_NAME_ _NEW_NAME_ ++
*sketchybar --remove* _NAME_

*sketchybar --add graph* _NAME_ _POSITION_ _WIDTH_ *--push* _NAME_
_DATA_POINT_ ... _DATA_POINT_ ++
*sketchybar --add space* _NAME_ _POSITION_ ++
*sketchybar --add bracket* _NAME_ _MEMBER_NAME_ ... _MEMBER_NAME_ ++
*sketchybar --add alias* _APPLICATION_NAME_|"_WINDOW_OWNER_,_WINDOW_NAME_" _POSITION_ ++
*sketchybar --add slider* _NAME_ _POSITION_ _WIDTH_

*sketchybar --set* _NAME_ *popup*._PROPERTY_=_VALUE_

*sketchybar --add event* _NAME_ [_NSDistributedNotificationName_] ++
*sketchybar --subscribe* _NAME_ _EVENT_ ... _EVENT_ ++
*sketchybar --trigger* _NAME_ [_ENVVAR_=_VALUE_ ... _ENVVAR_=_VALUE_] ++
*sketchybar --update*

*sketchybar --query* *bar*|_NAME_|*defaults*|*events*|*default_menu_items*

*sketchybar --animate* _CURVE_ _DURATION_ *--bar* _SETTING_=_VALUE_ ...
_SETTING_=_VALUE_ ++
*sketchybar --animate* _CURVE_ _DURATION_ *--set* _PROPERTY_=_VALUE_ ...
_PROPERTY_=_VALUE_

# OPTIONS

*--config* _FILE_
	Specify an alternative configuration file. The default one found at
	_$HOME/.config/sketchybar/sketchybarrc_ is automatically read when sketchybar
	is launched.

*--reload* [_CONFIG_FILE_]
	Reload the bar without manually restarting the process. Essentially, this
	has the same effect as restarting the process, but is a bit more convenient.
	Additionally, an optional path to a new _CONFIG_FILE_ can be given to load
	a different configuration. If the optional argument is left out, the current
	configuration is reloaded.

*--hotload* _BOOLEAN_
	Enables the hotload feature. If you wish that the bar automatically reloads
	the configuration file once it is edited, the hotload functionality can be
	used. It will monitor the directory of the current configuration for changes
	and reload the configuration should it detect file changes.

For the other commands, see CONFIGURATION below for the relevant man page.

# DESCRIPTION

*sketchybar* aims to create a highly flexible, customizable, fast and powerful
status bar replacement for people that like playing with shell scripts.

Its features include
- performance friendly
- fully scriptable
- fully configurable (fonts, backgrounds, colors, icons, etc.)
- support for drawing native macOS menu bar applications (aliases)
- powerful event and scripting system
- animation system
- popup menus
- mouse support
- support for graphs
- per display and per space individualization.

The _bar_ is a rectangle that can hold arbitrarily many items which can be
configured to do awesome stuff. An _item_ will occupy a space in the bar and can
be equipped to show an icon and a label. The _icon_ and _label_ can be changed
through _scripts_ that can be attached to the item. It is also possible to
_subscribe_ an item to certain events for their script execution action, which
makes very powerful items possible.

Furthermore, an item can be assigned to mission control spaces or displays,
such that they only show on a certain space or display. This makes
multi-desktop configuration of the bar possible. This also opens the possibility
to create individualized bar configuration on a per display and per space level.

These simple ingredients make items almost endlessly customizable and can
be used to display arbitrary information and perform useful actions. Some
special features can not be accomplished with a simple item. This is where the
_components_ come into play which basically are items with extra steps. They
contain all the properties a regular item does, but they can do specialized
tasks a simple item can not. For example, there is a graph component which can
be used to display graphs in the bar.

For more details on how the configuration works, see CONFIGURATION.

# CONFIGURATION

A sketchybar configuration file is a regular shell script file that resides
at _$HOME/.config/sketchybar/sketchybarrc_. It is executed when sketchybar
launches. Everything persistent should be set up in this file.

It is possible to play with properties in the command line and change them
on the fly while the bar is running. Once a fitting value is found, the
configuration can then be included in the configuration file. The configuration
is then restored on restart.

When configuring the bar, it can be helpful to *stop the brew service* and *run
sketchybar* from the command line directly to see all relevant error messages
and warnings directly.

Refer to
- *sketchybar*(1) (this man page) for the overview of sketchybar
- *sketchybar*(5) to configure the bar
- *sketchybar-items*(5) to configure items
- *sketchybar-components*(5) to configure special components
- *sketchybar-popup*(5) to configure popup menus
- *sketchybar-events*(5) to configure events and scripting
- *sketchybar-query*(5) to query information
- *sketchybar-animate*(5) to animate
- *sketchybar-types*(5) for the nomenclature of the value types used in sketchybar
- *sketchybar-tips*(5) for configuration tips and tricks.

# SETUP

## Running sketchybar

Run the bar automatically at startup:
```
brew services start sketchybar
```

or in the command line with verbose output:
```
sketchybar
```

## Fonts

The default sketchybar font is the Hack Nerd Font. To install:
```
brew tap homebrew/cask-fonts
brew install --cask font-hack-nerd-font
```

If there exists missing icons, you might need to install it. Any font of your
liking can be used in sketchybar.

If you want to load fonts from a non standard directory, you can use
```
sketchybar --load-font <path>
```
to load a font file from any <path>.

## Plugins

When additional plugins are used or created, make sure that they are made
executable via
```
chmod +x name_of_plugin.sh
```

If sketchybar is run from the command line directly with the command
*sketchybar*, all outputs and error messages from scripts will be printed in
stdout.

The default plugin folder is located at _$HOME/.config/sketchybar/plugins_.
Plugins need to be referenced with absolute paths because relative paths will
not be resolved correctly.

Have a look at the discussion for plugins and share your own if you want to.
Just make sure to vet the code from all plugins before executing them to make
sure they are not harming your computer.

## Hiding the system menu bar

To hide the original macOS bar in
- pre-Ventura: go to _System Preferences_ > _Dock & Menu Bar_, then check the
	  _Automatically hide and show the menu bar_ option
- Ventura: go to _System Settings_ > _Desktop & Dock_, then set the 
	  _Automatically hide and show the menu bar_ option to _Always_
- Sonoma: go to _System Settings_ > _Control Center_, then set the 
	  _Automatically hide and show the menu bar_ option to _Always_.

# SEE ALSO

*sketchybar*(5)++
*sketchybar-items*(5)++
*sketchybar-components*(5)++
*sketchybar-popup*(5)++
*sketchybar-events*(5)++
*sketchybar-query*(5)++
*sketchybar-animate*(5)++
*sketchybar-types*(5)++
*sketchybar-tips*(5)

# BUGS

Report them at https://github.com/FelixKratz/Sketchybar/issues.

# CREDITS

*Sketchybar* was forked from *spacebar* (see https://github.com/cmacrae/spacebar)
and completely reimagined and rewritten. The original idea is based on the
status bar that was included in *yabai* (see 
https://github.com/koekeishiya/yabai) before getting removed.

# MAINTAINERS

Felix Kratz <felix.kratz@tu-dortmund.de>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/cmacrae/spacebar/compare/v1.3.0...HEAD)

## [1.3.0](https://github.com/cmacrae/spacebar/releases/tag/v1.3.0) - 2020-04-25
**Changed**
- Fixed a bug introduced in Big Sur where buffer reads were incorrect
- Improved efficiency of bar's initialisation
- Fixed DoNotDisturb indicator on Big Sur

**Added**
- New `left|center|right` shell sections: display custom text based on shell pipelines
- Option to turn focused window title display on or off (thanks [@Norviah](https://github.com/Norviah)!)
- Option to turn the spaces indicator on or off
- Option to turn the clock on or off
- Option to turn the power indicator on or off
- Option to turn the DoNotDisturb indicator on or off
- Option to set the padding between the first/last item and the left/right edge of the display
- Default to "Solid" style of Font Awesome 5 Free icon font
- Option to display spaces for all displays, including: optional separator, secondary & tertiary space indicator colours in relation to display
- Option to specify only drawing one bar on the main display or multiple bars, one for each display (this aligns with yabai's `external_bar` option)
- Provide a [Nix Flake](https://nixos.wiki/wiki/Flakes)

## [1.2.1](https://github.com/cmacrae/spacebar/releases/tag/v1.2.1) - 2020-11-18

**Changed**
- Fixed a bug where querying for the value of `space_icon_strip` would set it to default (thanks [@jraregris](https://github.com/jraregris))

## [1.2.0](https://github.com/cmacrae/spacebar/releases/tag/v1.2.0) - 2020-11-18

**Added**
- Spacing configuration options

## [1.1.1](https://github.com/cmacrae/spacebar/releases/tag/v1.1.1) - 2020-07-21

**Changed**
- Fixed a bug where long window titles would draw over the right status area
- Padding between items in the status area based on current values

## [1.1.0](https://github.com/cmacrae/spacebar/releases/tag/v1.1.0) - 2020-07-17

**Added**
- Height configuration option

## [1.0.0](https://github.com/cmacrae/spacebar/releases/tag/v1.0.0) - 2020-07-16

**Added**
- Option to position at the top or bottom of the screen
- Individual colour settings for each icon in the right strip (`dnd`, `power`, `clock`)
- DoNotDisturb indicator

**Changed**
- Current space indicated by colouring the glyph
- Removal of underlines
- Fixed flicker bug when changing monitor focus (thanks [@tom-auger](https://github.com/tom-auger))

## Pre-1.0.0
This changelog was not kept up to date prior to `1.0.0`.  
See the commit log for more information.


src/sketchybar.c
#include "bar_manager.h"
#include "event.h"
#include "workspace.h"
#include "mach.h"
#include "mouse.h"
#include "message.h"
#include "power.h"
#include "wifi.h"
#include "misc/help.h"
#include "media.h"
#include "hotload.h"
#include <libgen.h>

#define LCFILE_PATH_FMT  "/tmp/%s_%s.lock"

#define CLIENT_OPT_LONG  "--message"
#define CLIENT_OPT_SHRT  "-m"

#define VERSION_OPT_LONG "--version"
#define VERSION_OPT_SHRT "-v"

#define CONFIG_OPT_LONG  "--config"
#define CONFIG_OPT_SHRT  "-c"

#define HELP_OPT_LONG    "--help"
#define HELP_OPT_SHRT    "-h"

#define MAJOR 2
#define MINOR 21
#define PATCH 0

#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 140000
extern CGError SLSWindowManagementBridgeSetDelegate(void* delegate);
#endif

extern CGError SLSRegisterNotifyProc(void* callback, uint32_t event, void* context);
extern int SLSGetSpaceManagementMode(int cid);
extern int SLSMainConnectionID(void);
extern int RunApplicationEventLoop(void);

int g_connection;
CFTypeRef g_transaction;

struct bar_manager g_bar_manager;
struct mach_server g_mach_server;
void *g_workspace_context;

char g_name[256];
char g_config_file[4096];
char g_lock_file[MAXLEN];
bool g_volume_events;
bool g_brightness_events;
int64_t g_disable_capture = 0;
pid_t g_pid = 0;

static int client_send_message(int argc, char **argv) {
  if (argc <= 1) {
    return EXIT_SUCCESS;
  }

  char *user = getenv("USER");
  if (!user) {
    error("sketchybar-msg: 'env USER' not set! abort..\n");
  }

  int message_length = argc;
  int argl[argc];

  for (int i = 1; i < argc; ++i) {
    argl[i] = strlen(argv[i]);
    message_length += argl[i] + 1;
  }

  char* message = malloc((sizeof(char) * (message_length + 1)));
  char* temp = message;

  for (int i = 1; i < argc; ++i) {
    memcpy(temp, argv[i], argl[i]);
    temp += argl[i];
    *temp++ = '\0';
  }
  *temp++ = '\0';

  char bs_name[256];
  snprintf(bs_name, 256, MACH_BS_NAME_FMT, g_name);

  char* rsp = mach_send_message(mach_get_bs_port(bs_name),
                                message,
                                message_length,
                                true                     );

  free(message);
  if (!rsp) return EXIT_SUCCESS;

  if (strlen(rsp) > 2 && rsp[1] == '!') {
    fprintf(stderr, "%s", rsp);
    return EXIT_FAILURE;
  } else {
    fprintf(stdout, "%s", rsp);
  }
  free(rsp);

  return EXIT_SUCCESS;
}

static void acquire_lockfile(void) {
  int handle = open(g_lock_file, O_CREAT | O_WRONLY, 0600);
  if (handle == -1) {
    error("%s: could not create lock-file! abort..\n", g_name);
  }

  struct flock lockfd = {
    .l_start  = 0,
    .l_len    = 0,
    .l_pid    = getpid(),
    .l_type   = F_WRLCK,
    .l_whence = SEEK_SET
  };

  if (fcntl(handle, F_SETLK, &lockfd) == -1) {
    error("%s: could not acquire lock-file... already running?\n", g_name);
  }
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
static inline void init_misc_settings(void) {
  char *user = getenv("USER");
  if (!user) {
    error("%s: 'env USER' not set! abort..\n", g_name);
  }

  snprintf(g_lock_file, sizeof(g_lock_file), LCFILE_PATH_FMT, g_name, user);

  signal(SIGCHLD, SIG_IGN);
  signal(SIGPIPE, SIG_IGN);
  CGSetLocalEventsSuppressionInterval(0.0f);
  CGEnableEventStateCombining(false);
  g_connection = SLSMainConnectionID();
  g_volume_events = false;
  g_brightness_events = false;
}
#pragma clang diagnostic pop

static void parse_arguments(int argc, char **argv) {
  if ((string_equals(argv[1], VERSION_OPT_LONG))
      || (string_equals(argv[1], VERSION_OPT_SHRT))) {
    fprintf(stdout, "sketchybar-v%d.%d.%d\n", MAJOR, MINOR, PATCH);
    exit(EXIT_SUCCESS);
  } else if ((string_equals(argv[1], HELP_OPT_LONG))
      || (string_equals(argv[1], HELP_OPT_SHRT))) {
    printf(help_str, argv[0]);
    exit(EXIT_SUCCESS);
  } else if ((string_equals(argv[1], CLIENT_OPT_LONG))
             || (string_equals(argv[1], CLIENT_OPT_SHRT))) {
    exit(client_send_message(argc-1, argv+1));
  } else if ((string_equals(argv[1], CONFIG_OPT_LONG))
             || (string_equals(argv[1], CONFIG_OPT_SHRT))) {
    if (argc < 3) {
      printf("[!] Error: Too few arguments for argument 'config'.\n");
    } else {
      if (set_config_file_path(argv[2])) return;
      printf("[!] Error: Specified config file path invalid.\n");
    }
    exit(EXIT_FAILURE);
  }

  exit(client_send_message(argc, argv));
}

static void space_events(uint32_t event, void* data, size_t data_length, void* context) {
  struct event ev = { NULL, SPACE_CHANGED };
  event_post(&ev);
}

static void system_events(uint32_t event, void* data, size_t data_length, void* context) {
  if (event == 1322) {
    g_disable_capture = clock_gettime_nsec_np(CLOCK_MONOTONIC_RAW_APPROX);
  } else if (event == 905) {
    g_disable_capture = -1;
  } else {
    g_disable_capture = 0;
  }
}

int main(int argc, char **argv) {
  snprintf(g_name, sizeof(g_name), "%s", basename(argv[0]));

  if (is_root())
    error("%s: running as root is not allowed! abort..\n", g_name);

  setenv("BAR_NAME", g_name, 1);

  if (argc > 1) parse_arguments(argc, argv);

  pid_for_task(mach_task_self(), &g_pid);
  init_misc_settings();
  acquire_lockfile();

  if (SLSGetSpaceManagementMode(g_connection) != 1) {
    error("%s: 'System Settings' -> 'Desktop & Dock' -> 'Displays have separate spaces' needs to be enabled for sketchybar to work properly.\n", g_name);
  }

  SLSRegisterNotifyProc((void*)system_events, 904, NULL);
  SLSRegisterNotifyProc((void*)system_events, 905, NULL);
  SLSRegisterNotifyProc((void*)system_events, 1401, NULL);
  SLSRegisterNotifyProc((void*)system_events, 1508, NULL);
  SLSRegisterNotifyProc((void*)system_events, 1322, NULL);

  if (__builtin_available(macOS 13.0, *)) {
    SLSRegisterNotifyProc((void*)space_events, 1327, NULL);
    SLSRegisterNotifyProc((void*)space_events, 1328, NULL);
  }

  struct event init = { NULL, INIT_MUTEX };
  event_post(&init);

  workspace_event_handler_init(&g_workspace_context);
  bar_manager_init(&g_bar_manager);

  mouse_begin();
  display_begin();
  workspace_event_handler_begin(&g_workspace_context);

  windows_freeze();
  bar_manager_begin(&g_bar_manager);
  windows_unfreeze();

  if (!mach_server_begin(&g_mach_server, mach_message_handler))
    error("%s: could not initialize daemon! abort..\n", g_name);

  begin_receiving_power_events();
  begin_receiving_network_events();
  initialize_media_events();

  exec_config_file();
  begin_receiving_config_change_events();

  #if __MAC_OS_X_VERSION_MAX_ALLOWED >= 140000
  if (__builtin_available(macos 14.0, *)) {
    SLSWindowManagementBridgeSetDelegate(NULL);
  }
  #endif

  RunApplicationEventLoop();
  return 0;
}
